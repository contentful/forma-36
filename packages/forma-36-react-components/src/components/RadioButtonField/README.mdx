---
title: 'RadioButtonField'
type: 'component'
status: 'stable'
slug: /components/radio-button-field/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/RadioButtonField'
storybook: 'https://f36-storybook.contentful.com/?path=/story/form-elements-radiobuttonfield--basic'
---

Use RadioButtonField to give users the possibility to select only one option from a number of choices. It contains a radio input with its label. If needed the component can display an error message and hint message. For a multiple-choice, please consider using [ChecboxField component](https://f36.contentful.com/components/checkbox-field/).

## Table of contents

- [Table of contents](#table-of-contents)
- [How to use RadioButtonField](#how-to-use-radiobuttonfield)
- [Code examples](#code-examples)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use RadioButtonField

- RadioButtonField should be used in a group. Please include at least two or more choices for users
- Options provided for users should be mutually exclusive.
- Make sure choices donâ€™t overlap. Avoid patterns like: 1990-2000, 2000-2010.
- Be explicit about the action that will follow if the radio button is selected

## Code examples

Basic usage

```jsx
// import {RadioButtonField} from '@contentful/f36-components';

<RadioButtonField label="Label text" id="randomId" />
```

Example with 2 options with helper text displayed

```jsx
// import {RadioButtonField} from '@contentful/f36-components';

function RadioButtonFieldExample() {
  const [activeOption, setActiveOption] = React.useState('yes');

  return (
    <Flex flexDirection="column">
      <RadioButtonField
        name="radio option 1"
        id="randomId1"
        value="yes"
        checked={activeOption === 'yes'}
        onChange={(e) => {
          setActiveOption(e.target.value);
        }}
        label="Radio button checked"
        helpText="Helper text displayed below"
      />
      <RadioButtonField
        name="radio option 2"
        id="randomId2"
        value="no"
        isChecked={activeOption === 'no'}
        onChange={(e) => {
          setActiveOption(e.target.value);
        }}
        label="Second option"
        helpText="Helper text displayed below"
      />
    </Flex>
  );
}
```

## Content recommendations

- Labels should be concise and provide context.
- Use sentence-style capitalization (only the first word in a phrase and any proper nouns capitalized)

## Accessibility

- Radio buttons should have a clear and concise label.
