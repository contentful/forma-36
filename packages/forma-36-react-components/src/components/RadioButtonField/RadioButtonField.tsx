import React from 'react';
import { ControlledInputField } from '@contentful/f36-forms';
import type { ControlledInputFieldProps } from '@contentful/f36-forms';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export type RadioButtonFieldProps = Omit<
  ControlledInputFieldProps,
  'inputType'
>;

export const RadioButtonField = ({
  isChecked = false,
  isLabelLight = false,
  testId = 'cf-ui-radio-button-field',
  ...otherProps
}: RadioButtonFieldProps) => {
  return (
    <ControlledInputField
      isChecked={isChecked}
      isLabelLight={isLabelLight}
      testId={testId}
      {...otherProps}
      inputType="radio"
    />
  );
};
