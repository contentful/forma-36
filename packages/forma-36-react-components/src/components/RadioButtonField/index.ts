export * from './RadioButtonField';
export { default } from './RadioButtonField';
