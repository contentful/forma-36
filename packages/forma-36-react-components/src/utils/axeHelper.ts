import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
