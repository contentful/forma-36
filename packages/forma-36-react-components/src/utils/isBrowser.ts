const isBrowser = () => typeof window !== 'undefined';

export default isBrowser;
