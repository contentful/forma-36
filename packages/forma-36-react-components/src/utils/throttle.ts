const throttle = (delay = 200, fn: Function) => {
  let lastCall = 0;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const throttleExec = (...args: any[]) => {
    const now = new Date().getTime();
    if (now - lastCall < delay) {
      return;
    }
    lastCall = now;
    return fn(...args); // eslint-disable-line
  };
  return throttleExec;
};

export default throttle;
