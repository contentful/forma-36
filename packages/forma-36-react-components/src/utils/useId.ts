import { useState, useLayoutEffect } from 'react';

/**
 * This util was inspired by Reach UI package called auto-id
 * https://github.com/reach/reach-ui/blob/86a046f54d53b6420e392b3fa56dd991d9d4e458/packages/auto-id/src/index.ts
 *
 * What is this hook all about?
 * Sometimes we need unique but sequential IDs.
 * One example is that some Accessibility standards rely on the id of some elements.
 * This hook will make sure that we generate a sequence of IDs when rendering a group of components.
 *
 * Out of scope for this solution
 * This solution will break in server-rendered apps because there will be a mismatch between the server id and the client id.
 * But this solution can be extended to cover that, a good example of how to solve that is ReachUI solution already mentioned on the top.
 */

let initialId = 0;

export const useId = () => {
  const [id, setId] = useState(initialId);

  useLayoutEffect(() => {
    setId(++initialId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return id;
};
