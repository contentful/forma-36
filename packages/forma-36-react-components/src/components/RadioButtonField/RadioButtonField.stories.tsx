import React, { useState } from 'react';

import RadioButtonField, { RadioButtonFieldProps } from './RadioButtonField';
import FieldGroup from '../Form/FieldGroup';
import Flex from '../Flex/Flex';
import SectionHeading from '../Typography/SectionHeading';

export default {
  title: 'Form Elements/RadioButtonField',
  component: RadioButtonField,
  parameters: {
    propTypes: [RadioButtonField['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const Basic = (args: RadioButtonFieldProps) => {
  const [activeOption, setActiveOption] = useState('yes');
  return (
    <FieldGroup>
      <RadioButtonField
        {...args}
        name="someOption1"
        checked={activeOption === 'yes'}
        value="yes"
        onChange={(e) => {
          setActiveOption((e.target as HTMLInputElement).value);
        }}
        id="termsCheckbox"
      />
      <RadioButtonField
        {...args}
        name="someOption2"
        value="no"
        checked={activeOption === 'no'}
        onChange={(e) => {
          setActiveOption((e.target as HTMLInputElement).value);
        }}
        id="termsCheckboxOption2"
      />
    </FieldGroup>
  );
};

Basic.args = {
  labelText: 'some label text',
  validationMessage: 'validation message',
  helpText: 'help text',
};

export const overview = () => (
  <>
    <Flex marginBottom="spacingS">
      <SectionHeading element="h3">Radio button field default</SectionHeading>
    </Flex>
    <Flex marginBottom="spacingS">
      <RadioButtonField
        labelText="Label text"
        helpText="This is a helptext"
        name="someOption"
        value="no"
        id="radioButton1"
      />
    </Flex>
    <Flex marginBottom="spacingS">
      <SectionHeading element="h3">
        Radio button field with validation message
      </SectionHeading>
    </Flex>
    <Flex marginBottom="spacingS">
      <RadioButtonField
        labelText="Label text"
        helpText="This is a helptext"
        validationMessage="validationMessage"
        name="someOption"
        value="no"
        id="radioButton2"
      />
    </Flex>
    <Flex marginBottom="spacingS">
      <SectionHeading element="h3">Radio button disabled</SectionHeading>
    </Flex>
    <Flex marginBottom="spacingS">
      <RadioButtonField
        labelText="Label text"
        helpText="This is a helptext"
        disabled
        name="someOption"
        value="no"
        id="radioButton3"
      />
    </Flex>
    <Flex marginBottom="spacingS">
      <SectionHeading element="h3">
        Radio button with light label
      </SectionHeading>
    </Flex>
    <Flex marginBottom="spacingS">
      <RadioButtonField
        labelText="Label text"
        helpText="This is a helptext"
        name="someOption"
        value="no"
        labelIsLight
        id="radioButton4"
      />
    </Flex>
  </>
);
