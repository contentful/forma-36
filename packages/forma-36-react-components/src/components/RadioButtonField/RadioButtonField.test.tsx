import React from 'react';
import { render } from '@testing-library/react';

import { axe } from '../../utils/axeHelper';
import { RadioButtonField } from './RadioButtonField';

it('renders the component', () => {
  const { container } = render(
    <RadioButtonField
      id="radio-button"
      inputProps={{ labelText: 'radio button field' }}
      labelText="radio button field"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(
    <RadioButtonField
      id="radio-button"
      className="my-extra-class"
      labelText="radio button field"
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(
    <RadioButtonField id="radio-button" labelText="radio button field" />,
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
