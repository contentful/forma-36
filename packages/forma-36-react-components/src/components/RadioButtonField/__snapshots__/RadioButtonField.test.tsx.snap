// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3 {
  margin-bottom: 0;
}

.emotion-0 {
  display: inline-block;
  font-size: 0.875rem;
  opacity: 0;
  position: absolute;
  margin: 0;
  width: 1rem;
  height: 1rem;
  z-index: 1;
  cursor: pointer;
}

.emotion-0:focus + label {
  outline: none;
  border: 2px solid #2E75D4;
  box-shadow: 0px 0px 0px 3px #84B9F5;
}

.emotion-0:checked + label {
  border-color: #2E75D4;
  background: #2E75D4;
}

.emotion-0:disabled + label {
  border-color: #c3cfd5;
  background: #c3cfd5;
}

.emotion-0:checked + label::before {
  background: #ffffff;
}

.emotion-0:checked:disabled + label::before {
  background: #6B7888;
}

.emotion-0:disabled + label::before {
  background: #b4c3ca;
}

.emotion-1 {
  background: #ffffff;
  width: 1rem;
  height: 1rem;
  border: 2px solid #c3cfd5;
  fill: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-sizing: border-box;
  border-radius: 50%;
}

.emotion-1:before {
  content: "";
  border-radius: 50%;
  background-color: #ffffff;
  width: 6px;
  height: 6px;
}

.emotion-2 {
  display: inline-block;
  position: relative;
  text-overflow: ellipsis;
  line-height: 1.5rem;
  margin: 0;
  width: 1rem;
  height: 1rem;
  min-height: 1rem;
  overflow: visible;
  margin-right: 0.5rem !important;
  margin-top: 0.25rem !important;
}

<div
  class="emotion-4"
  data-test-id="cf-ui-radio-button-field"
>
  <div
    class="emotion-2"
    data-test-id="cf-ui-controlled-input"
  >
    <input
      aria-label="radio button field"
      class="emotion-0"
      id="radio-button"
      type="radio"
      value=""
    />
    <label
      class="emotion-1"
      for="radio-button"
    />
  </div>
  <div>
    <label
      class="FormLabel__FormLabel___3d6zQ emotion-3"
      data-test-id="cf-ui-form-label"
      for="radio-button"
    >
      radio button field
    </label>
  </div>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3 {
  margin-bottom: 0;
}

.emotion-0 {
  display: inline-block;
  font-size: 0.875rem;
  opacity: 0;
  position: absolute;
  margin: 0;
  width: 1rem;
  height: 1rem;
  z-index: 1;
  cursor: pointer;
}

.emotion-0:focus + label {
  outline: none;
  border: 2px solid #2E75D4;
  box-shadow: 0px 0px 0px 3px #84B9F5;
}

.emotion-0:checked + label {
  border-color: #2E75D4;
  background: #2E75D4;
}

.emotion-0:disabled + label {
  border-color: #c3cfd5;
  background: #c3cfd5;
}

.emotion-0:checked + label::before {
  background: #ffffff;
}

.emotion-0:checked:disabled + label::before {
  background: #6B7888;
}

.emotion-0:disabled + label::before {
  background: #b4c3ca;
}

.emotion-1 {
  background: #ffffff;
  width: 1rem;
  height: 1rem;
  border: 2px solid #c3cfd5;
  fill: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-sizing: border-box;
  border-radius: 50%;
}

.emotion-1:before {
  content: "";
  border-radius: 50%;
  background-color: #ffffff;
  width: 6px;
  height: 6px;
}

.emotion-2 {
  display: inline-block;
  position: relative;
  text-overflow: ellipsis;
  line-height: 1.5rem;
  margin: 0;
  width: 1rem;
  height: 1rem;
  min-height: 1rem;
  overflow: visible;
  margin-right: 0.5rem !important;
  margin-top: 0.25rem !important;
}

<div
  class="emotion-4 my-extra-class"
  data-test-id="cf-ui-radio-button-field"
>
  <div
    class="emotion-2"
    data-test-id="cf-ui-controlled-input"
  >
    <input
      aria-label="radio button field"
      class="emotion-0"
      id="radio-button"
      type="radio"
      value=""
    />
    <label
      class="emotion-1"
      for="radio-button"
    />
  </div>
  <div>
    <label
      class="FormLabel__FormLabel___3d6zQ emotion-3"
      data-test-id="cf-ui-form-label"
      for="radio-button"
    >
      radio button field
    </label>
  </div>
</div>
`;
